src/Router.js

```diff
@@ -114,21 +114,35 @@ export default class Router {
    * @param {Context} ctx
    */
   __runHandlers(handlers, ctx, callback) {
-    let i = 0;
+    let index = 0;
 
     let next = () => {
       if (this.__dispatchId !== ctx.dispatchId) {
         return;
       }
-      let fn = handlers[i]
-      i++
-      // capture i in closure to not fuck
-      var j = i;
-
-      if (fn) {
-        fn(ctx, next)
-        if (callback && j === handlers.length && this.__dispatchId === ctx.dispatchId) {
-          callback();
+      let handlerFnOrArray = handlers[index]
+      index++
+      // capture handler index in closure since index could be modified by another thread
```

Comment:
Not blocking and just out of curiosity, but can you link me to an article on the risk of this? Would have missed this if I were writing it.

Reply:
note that on line 17 `index` is defined outside of the `next` function, then the next function is called recursively to process the route handlers.  You'll see on line 142 below the next serial route handler is executed.  If that handler is async, then the function will return and continue on line 143 where we check/compare `handlerIndex`.  If we were to check `index` instead we would be at risk for the async handler to have completed and called next() (which increments `index`) before we check the value.  Assigning to handlerIndex before making any async requests effectively 'freezes' a copy of the index value so that we don't have to worry about concurrent modification.  This article on mutual exclusion talks about more advanced ways to protect thread-shared data:
https://www.developer.com/lang/jscript/article.php/3592016